### PR TITLE
[mono][interp] Fix memory leaks for interpreted dynamic methods.

### PR DESCRIPTION
**Pull request by vargaz**
* Add a mempool to MonoDynamicMethod.
* Modify the intepreter code to allocate from this mempool when interpreting dynamic methods.

### CODE REVIEW

**Comment by vargaz** [comment]
Ref: https://github.com/dotnet/runtime/issues/87078

File: src/mono/mono/mini/interp/interp.c [hunk 1]
```diff
@@ -3438,12 +3461,24 @@ static void
...
 	mono_internal_hash_table_remove (&jit_mm->interp_code_hash, method);
+	if (imethod && jit_mm->interp_method_pointer_hash) {
+		if (imethod->jit_entry)
+			g_hash_table_remove (jit_mm->interp_method_pointer_hash, imethod->jit_entry);
+		if (imethod->llvmonly_unbox_entry)
+			g_hash_table_remove (jit_mm->interp_method_pointer_hash, imethod->llvmonly_unbox_entry);
+	}
 	jit_mm_unlock (jit_mm);
+
+	if (dmethod->mp) {
```
Comment (kg) [inline src/mono/mono/mini/interp/interp.c:3478]: if this is done outside the lock, is there a possible race condition where dmethod->mp will get used during destruction?

Comment (vargaz) [reply src/mono/mono/mini/interp/interp.c:3478]: The method is freed after the corresponding DynamicMethod is GCd, at that point no one should be executing the method any more.

**Review by BrzVlad — APPROVED** [review]

**Comment by BrzVlad** [comment]
I suspect this might have caused crashes like `Assertion at /home/vbrezae/Xamarin/repos/runtime/src/mono/mono/mini/interp/transform.c:8244, condition `td->new_code [reloc->offset + reloc->skip + 1] == 0xdead'`.

I am able to reproduce with moderate difficulty. I'm taking a look into it.

**Pull request merged**